src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -210,33 +209,92 @@ private void checkLayoutRecursive(MemoryLayout layout) {
+                        "not supported because a sequence of a padding layout is not allowed");
+            }
             checkLayoutRecursive(sl.elementLayout());
         }
     }
 
-    // check for trailing padding
-    private void checkGroupSize(GroupLayout gl, long maxUnpaddedOffset) {
+    // check elements are not all padding layouts and for trailing padding
+    static private void checkGroup(GroupLayout gl, long maxUnpaddedOffset) {
+        PaddingLayout preceedingPadding = null;
+        List<MemoryLayout> memberLayouts = gl.memberLayouts();
+        for (int i = 0; i < memberLayouts.size(); i++) {
+            MemoryLayout current = memberLayouts.get(i);
+            if (current instanceof PaddingLayout pl) {
+                if (preceedingPadding != null) {
+                    throw new IllegalArgumentException("The padding layout " + pl +
+                            " was preceded by another padding layout " + preceedingPadding +
+                            inMessage(gl));
+                }
+                preceedingPadding = pl;
+            } else {
+                if (preceedingPadding != null) {
+                    assertIsAlignedBy(gl, i, preceedingPadding, current);
+                    preceedingPadding = null;
+                }
+            }
+        }
+        // Check an optional trailing padding in a struct layout used to align any previous members
+        if (gl instanceof StructLayout && (gl.memberLayouts().size() > 1) &&
+                gl.memberLayouts().getLast() instanceof PaddingLayout pl) {
+            assertIsAlignedBy(null, 0, pl, gl);
+        }
+
+        if (!gl.memberLayouts().isEmpty() && gl.memberLayouts().stream().allMatch(e -> e instanceof PaddingLayout)) {
+            throw new IllegalArgumentException("Layout '" + gl + "' is non-empty and only has padding layouts");
+        }
         long expectedSize = Utils.alignUp(maxUnpaddedOffset, gl.byteAlignment());
         if (gl.byteSize() != expectedSize) {
             throw new IllegalArgumentException("Layout '" + gl + "' has unexpected size: "
                     + gl.byteSize() + " != " + expectedSize);
         }
     }
 
+    static void assertIsAlignedBy(GroupLayout gl, long index, PaddingLayout padding, MemoryLayout element) {
+        if (padding.byteSize() > element.byteAlignment()) {
+            throw new IllegalArgumentException("The padding layout " + padding +
+                    " is not of minimum size to align " + element +
+                    "(with byte alignment " + element.byteAlignment() + ")" + inMessage(gl));
+        }
+        long pos = (gl != null) ? gl.byteOffset(MemoryLayout.PathElement.groupElement(index)) : 0;
+        if (!Utils.isAligned(pos, element.byteAlignment())) {
+            throw new IllegalArgumentException("The padding layout " + padding +
+                    " does not align the element " + element +
+                    " (at byte offset " + pos + ")" + inMessage(gl));
+        }
```

Comment:
What happens here if an element is already aligned, and `padding.byteSize() == element.byteAlignment()`? e.g. `MemoryLayout.structLayout(MemoryLayout.paddingLayout(4), ValueLayout.JAVA_INT)`?

Reply:
Also, don't we already check the alignment before when calling `checkMemberOffset`? Why is it checked again here?

Reply:
> What happens here if an element is already aligned, and `padding.byteSize() == element.byteAlignment()`? e.g. `MemoryLayout.structLayout(MemoryLayout.paddingLayout(4), ValueLayout.JAVA_INT)`?

This is an error that existed before this PR. On the main branch:

```
    @Test
    public void alignedByInitialPadding() {
        Linker linker = Linker.nativeLinker();
        var struct = MemoryLayout.structLayout(
                MemoryLayout.paddingLayout(Integer.BYTES),
                JAVA_INT);
        var fd = FunctionDescriptor.of(struct, struct, struct);
        linker.downcallHandle(fd);
    }
```

would produce:

```
test TestLinker.alignedByInitialPadding(): failure
java.lang.IllegalArgumentException: Member layout 'i4', of '[x4i4]' found at unexpected offset: 4 != 0
	at java.base/jdk.internal.foreign.abi.AbstractLinker.checkMemberOffset(AbstractLinker.java:235)
	at java.base/jdk.internal.foreign.abi.AbstractLinker.checkLayoutRecursive(AbstractLinker.java:195)
	at java.base/jdk.internal.foreign.abi.AbstractLinker.checkLayout(AbstractLinker.java:176)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at java.base/jdk.internal.foreign.abi.AbstractLinker.checkLayouts(AbstractLinker.java:167)
	at java.base/jdk.internal.foreign.abi.AbstractLinker.downcallHandle0(AbstractLinker.java:98)
	at java.base/jdk.internal.foreign.abi.AbstractLinker.downcallHandle(AbstractLinker.java:92)
	at TestLinker.alignedByInitialPadding(TestLinker.java:160)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)

```

Reply:
I've raised a separate issue for this: https://bugs.openjdk.org/browse/JDK-8343620

Reply:
Looks like you removed the check here: https://github.com/openjdk/jdk/pull/21041/commits/0c134821d6ad7d3c7668d9cc88f7d329b5648827

I think that is good. We already check whether struct fields are aligned when creating a struct layout.

The existing `checkMemberOffset` does the right thing for this case. If tries to manually compute the expected offset of the field (using minimal amount of padding), and then checks if the field actually appears at that offset.